docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,186 @@
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) can't - the contents of the struct aren't known.
+
+From the Rust side of the world, this is probably solvable by sprinkling more macros around - eg, something like:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+Might be enough for the generation of the Rust scaffolding. However, the problems are in the foreign bindings.
+
+## How the type universe is constructed for the macro approach.
+
+In both diplomat and [#416](https://github.com/mozilla/uniffi-rs/pull/416), the approach taken
+is that the generation process wants a path to the Rust source file that contains the module in
+question - in the example above, the `ffi` module annotated with `#[diplomat:bridge]`. They both
+use the `syn` crate to parse the Rust code inside this module, build their type universe, then
+generate the foreign bindings.
+
+In our problematic example above, this process never sees the layout of the `MyFFIType` struct,
+and nor does it see any macros annotating them.
+
+For this approach to work, it would be necessary for this process to compile the entire crate,
+including depedent crates - the actual definition of all the types might appear anywhere.
+Not only would this be slow, it's not clear it could be made to work - it might be reasonable to
+have constraints on what can appear in just the `ffi` mod, but if we started adding constraints
+to the entire crate, the tool would become far less useful.
+
+This is the exact same problem which caused us to decide to stop working on
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) - the current world where the type universe
+is described externally doesn't have this problem - only the UDL file needs to be parsed when
+generating the foreign bindings - Rust code isn't considered. The application-services team has
+concluded that none of our non-trival use-cases for UniFFI could be described using macros,
+so supporting both mechanisms is pain for no gain.
+
+As noted in #416, `wasm-bindgen` has a similarly shaped problem, and solves it by having
+the Rust macro arrange for the resulting library to have an extra data section with the
+serialized "type universe" - foreign binding generation would then read this information from the
+already built binary. This sounds more complex than the UniFFI team has appetite for at
+the current time.
+
+## Is this a problem for users of diplomat? Will diplomat solve it?
+
+I couldn't find real examples using diplomat, so it's difficult to know if this
+is a problem in practice. UniFFI came from a world where we had Rust crates and
+a hand-written FFI that exposed types from all over the crate. If these tools
+had started with the limitations from the macro approach in mind, it's possible
+a different, acceptable design might have been made to work. Maybe duplicating
+some structs and supplying suitable `Into` implementations might make things workable?
```

Comment:
Yeah so we have a plan here: we will likely provide helpers that allow for duplicating structs and C-like enums with Into impls. Just haven't gotten around to it yet.

We do not want to actually allow for straight up using foreign non-opaque types, because it becomes a safety issue if these types ever change (e.g. someone removes the `#[repr(C)]`).

Come to think of it: how does uniffi handle this? Is there any way to expose types that are not defined by uniffi? It seems like y'all would have isomorphic problems.

Reply:
Ah, it seems like uniffi never directly exposes the rust values across ffi, it encodes them into its own format independent of `repr(C)` and `repr(Rust)`. clever.

Reply:
> . If Diplomat lets you duplicate the struct into the bridge module, and decorate it with a helper macro in order to generate some automatic Into helpers, then that feels pretty close to the same amount of work as what you have to do with UniFFI:

Note that we don't allow this just yet, but ... "close enough"; it's a very easy feature to add and we mostly haven't added it because we haven't needed it that much yet.

Reply:
I've tried to capture this too